### ISSUE TITLE
AutoComplete: forceSelection causes clear event to fire even when item selected

### ISSUE THREAD

**Issue by firsthour** [primefaces/primefaces#5731]
## 1) Environment
- PrimeFaces version: 8.0
- Does it work on the newest released PrimeFaces version? Version? No, 8.0
- Does it work on the newest sources in GitHub? (Build by source -> https://github.com/primefaces/primefaces/wiki/Building-From-Source) unknown
- Application server + version: Wildfly 18
- Affected browsers: Firefox 74

## 2) Expected behavior

With forceSelection=true and when both itemSelect and clear events are defined, enter text that matches part of an auto complete search, click one of the entries in the dropdown, and only the itemSelect event fires.

## 3) Actual behavior

The clear event is firing and then the itemSelect event fires.

## 4) Steps to reproduce

See code below. Enter some text in the p:autoComplete, pick any of the results, the clear will fire and then the select event will fire. This seems to be from #3270, where the clear event was never firing when forceSelection was true, but looking at the code, it's comparing what was typed in the p:autoComplete against what was selected, this feels wrong to me. This would require a user to type out the entire matching text for the clear event not to fire.

from autocomplete.js:

```
    isValid: function(value) {
        if(!this.cfg.forceSelection) {
            return;
        }

        var valid = false;

        for(var i = 0; i < this.currentItems.length; i++) {
            var stripedItem = this.currentItems[i];
            if (stripedItem) {
                stripedItem = stripedItem.replace(/\r?\n/g, '');
            }

            if(stripedItem === value) {
                valid = true;
                break;
            }
        }

        if(!valid) {
            this.input.val('');
            if(!this.cfg.multiple) {
                this.hinput.val('');
            }
            this.fireClearEvent();
        }

        return valid;
    },
```

We're migrating from 6.2.6 so this is new to us, we use forceSelection=true basically anywhere we use autoComplete so either we're doing something wrong and we're misusing forceSelection, or the source is too aggressive when matching. Thanks!

## 5) Sample XHTML

```
			<p:autoComplete
				placeholder="Search"
				value="#{testController.name}"
				forceSelection="true"
				completeMethod="#{testController.search}"
				var="name"
				itemLabel="#{name}"
				itemValue="#{name}">
				<p:ajax
					event="itemSelect"
					listener="#{testController.onSelect}" />
				<p:ajax
					event="clear"
					listener="#{testController.onClear}" />
			</p:autoComplete>
```

## 6) Sample bean

```
@Named
@ViewScoped
public class TestController implements Serializable {
	private static final long serialVersionUID = 1L;
	
	private String name;
	
	public List<String> search(String search) {
		return Arrays.asList("Greg", "Matt", "Aaron");
	}
	
	public void onSelect(SelectEvent<String> event) {
		System.out.println("onCustomerSelect()");
	}
	
	public void onClear(AjaxBehaviorEvent event) {
		System.out.println("onCustomerClear()");
	}
	
	public String getName() {
		return name;
	}
	
	public void setName(String name) {
		this.name = name;
	}
}
```

**Comment by melloware** [comment]
Hmm you said you are upgrading from 6.2.6 but that isValid check which you think is too aggressive was introduced in 6.2.1 with #3428 

Then  #3270 is a 1 line fix that addresses the issue reported in that ticket.

But you could be right this introduces a new issue due to the ordering.  @fanste any thoughts as you originally reported #3270.

**Comment by fanste** [comment]
This looks lile some kind of race condition again.. I've not done any tests yet, but from looking at the code I would say, that the input's blur event is the problem.

You enter something in the textfield than you click on an item in the dropdown panel. This leaves the textfield fireing the blur event before any click event is propagated from the panel.

**Comment by melloware** [comment]
OK I will do some investigation today.

**Comment by firsthour** [comment]
Sorry about the red herring about the autocomplete.js isValid() code.

Certainly seems like a race condition as you said. Clicking a dropdown fires the blur function in setupForceSelection() with the typed value, and then you can fire it again by clicking outside the input text with the selected value.

Unfortunately, it's not clear to me what would have changed from 6.2.6. I have confirmed in our 6.2.6 branch that the clear event does not fire on dropdown selection.

```
    setupForceSelection: function() {
        this.currentItems = [this.input.val()];
        var $this = this;

        this.input.blur(function() {
            var value = $(this).val(),
            valid = $this.isValid(value);

            if($this.cfg.autoSelection && valid && $this.checkMatchedItem && $this.items && !$this.isTabPressed && !$this.itemSelectedWithEnter) {
                var selectedItem = $this.items.filter('[data-item-label="' + $.escapeSelector(value) + '"]');
                if (selectedItem.length) {
                    selectedItem.click();
                }
            }

            $this.checkMatchedItem = false;
        });
    },
```

**Comment by melloware** [comment]
@firsthour Add this script to your page or your JS file that loads after PrimeFaces.
```javascript
PrimeFaces.widget.AutoComplete.prototype.deleteClearTimeout = function () {
   clearTimeout(this.clearingTimeout);
   this.clearingTimeout = null;
};

PrimeFaces.widget.AutoComplete.prototype.fireClearEvent = function () {
   var $this = this;
   $this.clearingTimeout = setTimeout(function () {
      $this.clearingTimeout = null;
      $this.callBehavior('clear');
   }, 100);
};

PrimeFaces.widget.AutoComplete.prototype.invokeItemSelectBehavior = function (event, itemValue) {
   if (this.clearingTimeout) {
      this.deleteClearTimeout();
   }

   if (this.hasBehavior('itemSelect')) {
      var ext = {
         params: [{
            name: this.id + '_itemSelect',
            value: itemValue
         }]
      };

      this.callBehavior('itemSelect', ext);
   }
};
```

**Comment by melloware** [comment]
I regression tested #3720 with the above fix and it remains fixed as well.

**Comment by fanste** [comment]
There might be a better solution than a timeout, which always feels a bit ugly.

https://developer.mozilla.org/en-US/docs/Web/API/FocusEvent/relatedTarget

The blur event may tell uns which element receives the focus. If we check if it is the panel, we could skip the clear event

**Comment by melloware** [comment]
@fanste let me give that a shot

**Comment by melloware** [comment]
OK I added this code
```javascript
 this.input.on('blur', function(event) {
            console.log(event.originalEvent.relatedTarget);
```

And the relatedTarget is ALWAYS NULL.  So when I use the mouse to select an item this relatedTarget is null. So I can't tell apart a good blur and a bad blur.  So we may have to stick with my timeout solution.

**Comment by firsthour** [comment]
Thanks, that javascript seems to work, though I would generally agree on avoiding alerts if possible, but I understand it's a necessary evil at times.

Also I did a diff on 6.2.6 and 8.0 and while isValid() is mostly unchanged,  `this.fireClearEvent();` was added when it's not valid, so that's the trigger of this change. Is it worth revisiting #3270? I don't know.

Edit: sorry if I'm like 24 hours behind on understanding the root cause of this ticket. :)

**Comment by melloware** [comment]
well if you look at #3270 there are 3 use cases listed there where the Clear event was not being sent when it should have.  If we remove the 1 line fix for clearEvent() then 2 of those use cases then fail now.

**Comment by fanste** [comment]
> 
> 
> OK I added this code
> 
> ```js
>  this.input.on('blur', function(event) {
>             console.log(event.originalEvent.relatedTarget);
> ```
> 
> And the relatedTarget is ALWAYS NULL. So when I use the mouse to select an item this relatedTarget is null. So I can't tell apart a good blur and a bad blur. So we may have to stick with my timeout solution.

The property `relatedTarget` is only set if the switched-to element is really focusable. A `div` can't receive focus until it has e.g. `tabindex="-1"`.

https://jsfiddle.net/71btm3kw/

**Comment by melloware** [comment]
Your fix works great I closed the old PR and opened a new one.

**Comment by fanste** [comment]
Thanks for your time to test it in PF. I'm currently a bit short on time :(

Anonther consideration: What happens with `forceSelection="true"` and `dropdownMode="current"`?

**Comment by melloware** [comment]
let me try that... will report back.

**Comment by melloware** [comment]
with `dropdownMode="current"` seems to be working the same correct way.

**Comment by PraveenTrex** [comment]
In which version of primefaces its fixed ? Currently I'm using 8.0 facing this issue @melloware @fanste

**Comment by tandraschko** [comment]
You can see the Milestone assigned to this issue

**Comment by PraveenTrex** [comment]
> You can see the Milestone assigned to this issue

Thanks @tandraschko 

Is there any other way to fix this issue in version 8.0 ? because we just updated our project from the version 6.2

**Comment by melloware** [comment]
@PraveenTrex nope not easily sine it requires both Java and Javascript changes there is no easy MonkeyPatch for this.  You would basically have to implement this yourself.  And why did you upgrade to 8.0 and not 11.0?   Going from 6.2 to 8.0 you may as well have gone to 11.0 which has over 1000 issues fixed since 8.0.

**Comment by PraveenTrex** [comment]
Thanks @melloware, I will update to version 11.0
